File: appliance/numerals.py
"""Numeral spelling subsystem.

Constraint Two of the appliance specification requires that every operator
facing surface renders numbers as words rather than as digit characters.  This
module is the single server side authority for that transformation.  A browser
side twin lives at ``web/js/numerals.js`` and is required to agree with this
module for every value the product can produce; ``tests/test_numerals.py``
enforces that agreement.

The public surface is deliberately small:

``spell_integer``   -- an exact integer rendered as words.
``spell_decimal``   -- a real number rendered as words, fraction digit by digit.
``spell_ordinal``   -- position words such as ``first`` or ``twenty-third``.
``spell_duration``  -- a second count rendered as a spoken duration.
``sanitize``        -- replace every digit run in arbitrary text with words.
``contains_digit``  -- the assertion used by the constraint tests.
"""

from __future__ import annotations

import re

__all__ = [
    "spell_integer",
    "spell_decimal",
    "spell_ordinal",
    "spell_duration",
    "sanitize",
    "contains_digit",
    "SpellingError",
]


class SpellingError(ValueError):
    """Raised when a value is outside the range this module can render."""


_SMALL = (
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight",
    "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen",
    "sixteen", "seventeen", "eighteen", "nineteen",
)

_TENS = (
    "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy",
    "eighty", "ninety",
)

# Short scale naming.  The empty first entry is the units group.
_SCALES = (
    "", "thousand", "million", "billion", "trillion", "quadrillion",
    "quintillion", "sextillion",
)

_GROUP = 1000

# Largest magnitude representable with the scale table above.
_LIMIT = _GROUP ** len(_SCALES)

_ORDINAL_SMALL = {
    "zero": "zeroth", "one": "first", "two": "second", "three": "third",
    "four": "fourth", "five": "fifth", "six": "sixth", "seven": "seventh",
    "eight": "eighth", "nine": "ninth", "ten": "tenth", "eleven": "eleventh",
    "twelve": "twelfth", "thirteen": "thirteenth", "fourteen": "fourteenth",
    "fifteen": "fifteenth", "sixteen": "sixteenth",
    "seventeen": "seventeenth", "eighteen": "eighteenth",
    "nineteen": "nineteenth", "twenty": "twentieth", "thirty": "thirtieth",
    "forty": "fortieth", "fifty": "fiftieth", "sixty": "sixtieth",
    "seventy": "seventieth", "eighty": "eightieth", "ninety": "ninetieth",
    "hundred": "hundredth", "thousand": "thousandth", "million": "millionth",
    "billion": "billionth", "trillion": "trillionth",
    "quadrillion": "quadrillionth", "quintillion": "quintillionth",
    "sextillion": "sextillionth",
}

_DIGIT_RUN = re.compile(r"\d+")
_ANY_DIGIT = re.compile(r"\d")
_WORD_EDGE = re.compile(r"[A-Za-z]")


def _spell_group(value: int) -> list[str]:
    """Render a value below one thousand as a list of words.

    An input of zero renders as an empty list so that group assembly can skip
    empty groups; the caller is responsible for the standalone zero case.
    """
    words: list[str] = []
    if value >= 100:
        words.append(_SMALL[value // 100])
        words.append("hundred")
        value %= 100
    if value >= 20:
        tens_word = _TENS[value // 10]
        remainder = value % 10
        if remainder:
            words.append(f"{tens_word}-{_SMALL[remainder]}")
        else:
            words.append(tens_word)
    elif value > 0:
        words.append(_SMALL[value])
    return words


def spell_integer(value: int) -> str:
    """Render an exact integer as words.

    Negative values are prefixed with ``negative``.  Grouping follows the short
    scale with no comma and no conjunction, so one thousand and twenty four
    renders as ``one thousand twenty-four``.
    """
    if isinstance(value, bool) or not isinstance(value, int):
        raise SpellingError(f"an exact integer is required, received {type(value).__name__}")

    if value == 0:
        return _SMALL[0]

    negative = value < 0
    magnitude = -value if negative else value

    if magnitude >= _LIMIT:
        raise SpellingError("the value exceeds the largest representable magnitude")

    # Split into groups of three, least significant group first.
    groups: list[int] = []
    while magnitude:
        groups.append(magnitude % _GROUP)
        magnitude //= _GROUP

    words: list[str] = []
    for index in range(len(groups) - 1, -1, -1):
        group_value = groups[index]
        if group_value == 0:
            continue
        words.extend(_spell_group(group_value))
        if index:
            words.append(_SCALES[index])

    rendered = " ".join(words)
    return f"negative {rendered}" if negative else rendered


def spell_decimal(value: float, places: int = 2) -> str:
    """Render a real number as words.

    The integral part is rendered as a whole number and the fractional part is
    rendered digit by digit after the word ``point``, which is how a number is
    read aloud.  A value whose fractional part rounds away renders as a plain
    whole number rather than trailing a run of zero words.
    """
    if isinstance(value, bool):
        raise SpellingError("a real number is required, received a boolean")
    if not isinstance(value, (int, float)):
        raise SpellingError(f"a real number is required, received {type(value).__name__}")
    if places < 0:
        raise SpellingError("the requested place count cannot be negative")
    if value != value or value in (float("inf"), float("-inf")):
        raise SpellingError("the value is not a finite real number")

    negative = value < 0
    text = f"{abs(float(value)):.{places}f}"
    whole_text, _, fraction_text = text.partition(".")

    fraction_text = fraction_text.rstrip("0")
    words = spell_integer(int(whole_text))
    if fraction_text:
        spoken = " ".join(_SMALL[int(character)] for character in fraction_text)
        words = f"{words} point {spoken}"
    return f"negative {words}" if negative and (int(whole_text) or fraction_text) else words


def spell_ordinal(value: int) -> str:
    """Render a position as words, such as ``first`` or ``one hundredth``."""
    cardinal = spell_integer(value)
    head, _, tail = cardinal.rpartition(" ")

    if "-" in tail:
        tens_part, _, units_part = tail.partition("-")
        tail = f"{tens_part}-{_ORDINAL_SMALL[units_part]}"
    else:
        tail = _ORDINAL_SMALL[tail]

    return f"{head} {tail}" if head else tail


def spell_duration(seconds: int) -> str:
    """Render a second count as a spoken duration.

    Zero renders as ``zero seconds``.  Units that are empty are omitted, so an
    exact hour renders as ``one hour`` rather than naming empty minutes and
    seconds.
    """
    if isinstance(seconds, bool) or not isinstance(seconds, int):
        raise SpellingError("an exact second count is required")
    if seconds < 0:
        raise SpellingError("a duration cannot be negative")
    if seconds == 0:
        return "zero seconds"

    remaining = seconds
    parts: list[str] = []
    for unit_seconds, singular, plural in (
        (86400, "day", "days"),
        (3600, "hour", "hours"),
        (60, "minute", "minutes"),
        (1, "second", "seconds"),
    ):
        count = remaining // unit_seconds
        remaining %= unit_seconds
        if count:
            name = singular if count == 1 else plural
            parts.append(f"{spell_integer(count)} {name}")
    return " ".join(parts)


def sanitize(text: str) -> str:
    """Replace every run of digit characters in ``text`` with its words.

    Each maximal run of digits is rendered independently as a whole number, so
    a dotted address renders as words separated by the original punctuation.
    Word boundaries are preserved: a space is introduced where a replacement
    would otherwise be welded onto an adjacent letter.

    This is the function the logging formatter applies to every emitted line,
    which is what makes Constraint Two unconditional rather than a convention
    that individual call sites are trusted to remember.
    """
    if not isinstance(text, str):
        text = str(text)
    if not _ANY_DIGIT.search(text):
        return text

    pieces: list[str] = []
    cursor = 0
    for match in _DIGIT_RUN.finditer(text):
        start, end = match.span()
        pieces.append(text[cursor:start])

        run = match.group()
        # Leading zeros carry meaning in identifiers, so they are spoken.
        stripped = run.lstrip("0")
        if stripped:
            leading = ["zero"] * (len(run) - len(stripped))
            try:
                spelled = " ".join(leading + [spell_integer(int(stripped))])
            except SpellingError:
                # Beyond the scale table, fall back to digit by digit.
                spelled = " ".join(_SMALL[int(character)] for character in run)
        else:
            spelled = " ".join(["zero"] * len(run))

        # Keep the replacement from fusing with adjacent letters.
        if pieces and pieces[-1] and _WORD_EDGE.search(pieces[-1][-1]):
            spelled = " " + spelled
        if end < len(text) and _WORD_EDGE.search(text[end]):
            spelled = spelled + " "

        pieces.append(spelled)
        cursor = end

    pieces.append(text[cursor:])
    return "".join(pieces)


def contains_digit(text: str) -> bool:
    """Report whether any digit character survives in ``text``."""
    return bool(_ANY_DIGIT.search(text))

File: web/js/numerals.js
/*
 * Numeral spelling subsystem, browser side.
 *
 * This is the twin of appliance/numerals.py and is required to produce
 * identical output for identical input.  The agreement is enforced by
 * tests/test_numerals.py, which executes this file and compares the two
 * implementations across the full range the product can produce.
 *
 * Written as a plain script with no module system and no build step, because
 * the dashboard is served as vanilla markup and scripting from an appliance
 * that may never see a package index.
 */

(function (root) {
    'use strict';

    var SMALL = [
        'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven',
        'eight', 'nine', 'ten', 'eleven', 'twelve', 'thirteen', 'fourteen',
        'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
    ];

    var TENS = [
        '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy',
        'eighty', 'ninety'
    ];

    var SCALES = [
        '', 'thousand', 'million', 'billion', 'trillion', 'quadrillion',
        'quintillion', 'sextillion'
    ];

    var ORDINAL_SMALL = {
        'zero': 'zeroth', 'one': 'first', 'two': 'second', 'three': 'third',
        'four': 'fourth', 'five': 'fifth', 'six': 'sixth', 'seven': 'seventh',
        'eight': 'eighth', 'nine': 'ninth', 'ten': 'tenth',
        'eleven': 'eleventh', 'twelve': 'twelfth', 'thirteen': 'thirteenth',
        'fourteen': 'fourteenth', 'fifteen': 'fifteenth',
        'sixteen': 'sixteenth', 'seventeen': 'seventeenth',
        'eighteen': 'eighteenth', 'nineteen': 'nineteenth',
        'twenty': 'twentieth', 'thirty': 'thirtieth', 'forty': 'fortieth',
        'fifty': 'fiftieth', 'sixty': 'sixtieth', 'seventy': 'seventieth',
        'eighty': 'eightieth', 'ninety': 'ninetieth', 'hundred': 'hundredth',
        'thousand': 'thousandth', 'million': 'millionth',
        'billion': 'billionth', 'trillion': 'trillionth',
        'quadrillion': 'quadrillionth', 'quintillion': 'quintillionth',
        'sextillion': 'sextillionth'
    };

    var GROUP = 1000;

    function spellGroup(value) {
        var words = [];
        if (value >= 100) {
            words.push(SMALL[Math.floor(value / 100)]);
            words.push('hundred');
            value = value % 100;
        }
        if (value >= 20) {
            var tensWord = TENS[Math.floor(value / 10)];
            var remainder = value % 10;
            words.push(remainder ? tensWord + '-' + SMALL[remainder] : tensWord);
        } else if (value > 0) {
            words.push(SMALL[value]);
        }
        return words;
    }

    function spellInteger(value) {
        if (typeof value !== 'number' || !isFinite(value) || Math.floor(value) !== value) {
            throw new Error('an exact integer is required');
        }
        if (value === 0) {
            return SMALL[0];
        }

        var negative = value < 0;
        var magnitude = Math.abs(value);

        // Grouping uses integer arithmetic so that values stay exact within
        // the safe integer range the dashboard actually produces.
        var groups = [];
        while (magnitude > 0) {
            groups.push(magnitude % GROUP);
            magnitude = Math.floor(magnitude / GROUP);
        }
        if (groups.length > SCALES.length) {
            throw new Error('the value exceeds the largest representable magnitude');
        }

        var words = [];
        for (var index = groups.length - 1; index >= 0; index -= 1) {
            if (groups[index] === 0) {
                continue;
            }
            words = words.concat(spellGroup(groups[index]));
            if (index) {
                words.push(SCALES[index]);
            }
        }

        var rendered = words.join(' ');
        return negative ? 'negative ' + rendered : rendered;
    }

    function spellDecimal(value, places) {
        if (places === undefined) {
            places = 2;
        }
        if (typeof value !== 'number' || !isFinite(value)) {
            throw new Error('a finite real number is required');
        }

        var negative = value < 0;
        var text = Math.abs(value).toFixed(places);
        var split = text.split('.');
        var wholeText = split[0];
        var fractionText = (split[1] || '').replace(/0+$/, '');

        var words = spellInteger(parseInt(wholeText, 10));
        if (fractionText) {
            var spoken = [];
            for (var index = 0; index < fractionText.length; index += 1) {
                spoken.push(SMALL[parseInt(fractionText.charAt(index), 10)]);
            }
            words = words + ' point ' + spoken.join(' ');
        }
        if (negative && (parseInt(wholeText, 10) !== 0 || fractionText)) {
            return 'negative ' + words;
        }
        return words;
    }

    function spellOrdinal(value) {
        var cardinal = spellInteger(value);
        var boundary = cardinal.lastIndexOf(' ');
        var head = boundary === -1 ? '' : cardinal.slice(0, boundary);
        var tail = boundary === -1 ? cardinal : cardinal.slice(boundary + 1);

        if (tail.indexOf('-') !== -1) {
            var halves = tail.split('-');
            tail = halves[0] + '-' + ORDINAL_SMALL[halves[1]];
        } else {
            tail = ORDINAL_SMALL[tail];
        }
        return head ? head + ' ' + tail : tail;
    }

    function spellDuration(seconds) {
        if (typeof seconds !== 'number' || Math.floor(seconds) !== seconds) {
            throw new Error('an exact second count is required');
        }
        if (seconds < 0) {
            throw new Error('a duration cannot be negative');
        }
        if (seconds === 0) {
            return 'zero seconds';
        }

        var units = [
            [86400, 'day', 'days'],
            [3600, 'hour', 'hours'],
            [60, 'minute', 'minutes'],
            [1, 'second', 'seconds']
        ];

        var remaining = seconds;
        var parts = [];
        for (var index = 0; index < units.length; index += 1) {
            var unitSeconds = units[index][0];
            var count = Math.floor(remaining / unitSeconds);
            remaining = remaining % unitSeconds;
            if (count) {
                var name = count === 1 ? units[index][1] : units[index][2];
                parts.push(spellInteger(count) + ' ' + name);
            }
        }
        return parts.join(' ');
    }


    /* Shapes that are identifiers rather than quantities, and are left alone.
     *
     * The rule is the difference between a number an operator reads and a
     * number an operator uses.  "Twelve active calls" is a quantity and reads
     * better spelled.  An address, a port, a version, a device name or a
     * response code is an identifier: the operator types it, compares it
     * against a label on a cable, or searches for it, and spelling it does not
     * make it clearer, it makes it unusable.
     *
     * This list is held identical to the one in the control plane's numeral
     * module, and a test asserts the two agree for every value the product can
     * produce. */
    var IDENTIFIER_SHAPES = [
        /\b(?:\d{1,3}\.){3}\d{1,3}\/\d{1,2}\b/g,
        /\b(?:\d{1,3}\.){3}\d{1,3}:\d{1,5}\b/g,
        /\b(?:\d{1,3}\.){3}\d{1,3}\b/g,
        /\b(?:[0-9a-fA-F]{0,4}:){2,7}[0-9a-fA-F]{0,4}\b/g,
        /\b(?:[0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}\b/g,
        /\b[0-9a-fA-F]{4}:[0-9a-fA-F]{2}:[0-9a-fA-F]{2}\.[0-9a-fA-F]\b/g,
        /\b[0-9a-fA-F]{4}:[0-9a-fA-F]{4}\b/g,
        /\b\d{4}-\d{2}-\d{2}(?:[T ]\d{2}:\d{2}(?::\d{2})?)?(?!\d)/g,
        /\b\d{2}:\d{2}(?::\d{2})?(?!\d)/g,
        /\bv?\d+\.\d+(?:\.\d+)*(?:[-+][0-9A-Za-z.]+)?\b/g,
        /\b(?:eth|en[a-z0-9]*|wl[a-z0-9]*|lo|tty[A-Za-z]*|sd[a-z]|nvme|dahdi|span|zap)\d+\b/g,
        /(?:\/[A-Za-z0-9._-]*\d[A-Za-z0-9._-]*)+/g,
        /\b(?:SIP|HTTP|status|code|error)\s+\d{3}\b/gi,
        /\b(?:chmod|mode|permissions)\s+[0-7]{3,4}\b/gi,
        /\b(?:errno|error\s+number)\s+\d+\b/gi,
        /\(\s*'[^']*'\s*,\s*\d{1,5}\s*\)/g,
        /\bport(?:\s+number)?\s+\d{1,5}\b/gi,
        /\b(?:TDM|TE|AEX|HA|HB|B)\d+[A-Z]?\b/g,
        /\bextension\s+\d+\b/gi
    ];

    /* A marker that cannot occur in real text, holding an identifier's place
     * while the quantities around it are spelled.  The index inside it is
     * written in letters rather than digits, because the marker passes through
     * the very function that turns digits into words and a numeric index would
     * be spelled along with everything else. */
    var GUARD_OPEN = '\uE000';
    var GUARD_CLOSE = '\uE001';
    var GUARD_FINDER = /\uE000([a-z]+)\uE001/g;

    function guardLabel(index) {
        var label = '';
        var remaining = index + 1;
        while (remaining > 0) {
            var remainder = (remaining - 1) % 26;
            remaining = Math.floor((remaining - 1) / 26);
            label = String.fromCharCode(97 + remainder) + label;
        }
        return label;
    }

    function guardIndex(label) {
        var index = 0;
        for (var position = 0; position < label.length; position += 1) {
            index = index * 26 + (label.charCodeAt(position) - 96);
        }
        return index - 1;
    }

    function identifierSpans(text) {
        var found = [];
        var shape;
        var match;
        for (var i = 0; i < IDENTIFIER_SHAPES.length; i += 1) {
            shape = IDENTIFIER_SHAPES[i];
            shape.lastIndex = 0;
            while ((match = shape.exec(text)) !== null) {
                if (match[0].length === 0) {
                    shape.lastIndex += 1;
                    continue;
                }
                found.push([match.index, match.index + match[0].length]);
            }
        }
        found.sort(function (left, right) {
            return left[0] - right[0] || (right[1] - right[0]) - (left[1] - left[0]);
        });
        var merged = [];
        for (var j = 0; j < found.length; j += 1) {
            if (merged.length && found[j][0] < merged[merged.length - 1][1]) {
                if (found[j][1] > merged[merged.length - 1][1]) {
                    merged[merged.length - 1][1] = found[j][1];
                }
                continue;
            }
            merged.push([found[j][0], found[j][1]]);
        }
        return merged;
    }

    function sanitize(text, spellIdentifiers) {
        if (typeof text !== 'string') {
            text = String(text);
        }
        if (!/[0-9]/.test(text)) {
            return text;
        }
        if (!spellIdentifiers) {
            var spans = identifierSpans(text);
            if (spans.length) {
                var held = [];
                var rebuilt = [];
                var walked = 0;
                for (var s = 0; s < spans.length; s += 1) {
                    rebuilt.push(text.slice(walked, spans[s][0]));
                    rebuilt.push(GUARD_OPEN + guardLabel(held.length) + GUARD_CLOSE);
                    held.push(text.slice(spans[s][0], spans[s][1]));
                    walked = spans[s][1];
                }
                rebuilt.push(text.slice(walked));
                var spelledText = sanitize(rebuilt.join(''), true);
                GUARD_FINDER.lastIndex = 0;
                return spelledText.replace(GUARD_FINDER, function (whole, label) {
                    return held[guardIndex(label)];
                });
            }
        }

        var pieces = [];
        var cursor = 0;
        var pattern = /[0-9]+/g;
        var match;

        while ((match = pattern.exec(text)) !== null) {
            var start = match.index;
            var end = start + match[0].length;
            pieces.push(text.slice(cursor, start));

            var run = match[0];
            var stripped = run.replace(/^0+/, '');
            var spelled;
            if (stripped) {
                var leading = [];
                for (var index = 0; index < run.length - stripped.length; index += 1) {
                    leading.push('zero');
                }
                try {
                    spelled = leading.concat([spellInteger(parseInt(stripped, 10))]).join(' ');
                } catch (error) {
                    var digits = [];
                    for (var position = 0; position < run.length; position += 1) {
                        digits.push(SMALL[parseInt(run.charAt(position), 10)]);
                    }
                    spelled = digits.join(' ');
                }
            } else {
                var zeros = [];
                for (var zeroIndex = 0; zeroIndex < run.length; zeroIndex += 1) {
                    zeros.push('zero');
                }
                spelled = zeros.join(' ');
            }

            var previous = pieces[pieces.length - 1];
            if (previous && /[A-Za-z]$/.test(previous)) {
                spelled = ' ' + spelled;
            }
            if (end < text.length && /[A-Za-z]/.test(text.charAt(end))) {
                spelled = spelled + ' ';
            }

            pieces.push(spelled);
            cursor = end;
        }

        pieces.push(text.slice(cursor));
        return pieces.join('');
    }

    function containsDigit(text) {
        return /[0-9]/.test(String(text));
    }

    var ApplianceNumerals = {
        spellInteger: spellInteger,
        spellDecimal: spellDecimal,
        spellOrdinal: spellOrdinal,
        spellDuration: spellDuration,
        sanitize: sanitize,
        containsDigit: containsDigit
    };

    /* The dashboard reads this from the global scope, and the cross
     * implementation agreement test loads the very same file through the
     * module system.  Both are served explicitly rather than by leaking a
     * declaration into whatever scope happens to be current. */
    root.ApplianceNumerals = ApplianceNumerals;
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = ApplianceNumerals;
    }
}(typeof globalThis !== 'undefined' ? globalThis : this));
